### PR TITLE
Updating the Code of this repo to support the new Meson build system

### PR DESCRIPTION
**Pull request by yozoon**
Hi! I know the project hit a road block a few months ago after you discovered that the images captured by this sensor are too small for the current matching algorithms. Nonetheless I thought it might be a good idea to update the codebase to match the changes of the freedesktop/libfprint upstream master branch so that if someone comes up with a clever idea to upscale the images (or something similar) we aren't too far behind the master branch. There were some substantial changes in the upstream master branch since the last commit (e.g. change of the build system) - so I decided to start from there and re-implement your changes in accordance with the new driver API.
As of now the driver is still not completely functional. I get the following error while waiting for the response of _vfs7552_image_start_:
`libusb: error [submit_bulk_transfer] submiturb failed error -1 errno=22`
But until this stage the driver seems to be communicating properly with the sensor.

This pull request is actually more intended to be a conversation starter - as I do not expect the code to be merged at this stage because it still is not en par with the functionality of your code. I currently don't know how to solve this error I encountered, so hopefully someone with a little more insight into the subject can investigate it further.
Cheers,
Julius

### CODE REVIEW

**Comment by hmaarrfk** [comment]
should i rebase this fork with upstream first? would my additions still work?

**Comment by yozoon** [comment]
As my code is already based on the freedesktop upstream there is no need to rebase it. In my fork I merged your driver code with the upstream master and fixed all compiler errors and warnings until the driver finally started talking to the device. So all your additions are actually in place - I just had to modify parts of them in order to accommodate it to the upstream changes.

File: libfprint/drivers/vfs7552_testing.c [hunk 1]
```diff
@@ -0,0 +1,873 @@
+/*
+ * Validity Sensors, Inc. VFS7552 Fingerprint Reader driver for libfprint
+ *                        ID 138a:0091 Validity Sensors, Inc.
+ * Copyright (C) 2013 Mark Harfouche <mark.harfouche@gmail.com>
```
Comment (hmaarrfk) [inline libfprint/drivers/vfs7552_testing.c]: 2013?

**Comment by hmaarrfk** [comment]
It is just really hard to focus on your contributions when 17k lines changed....

File: libfprint/drivers/vfs7552_testing.c (hunk 1 above)
Comment (yozoon) [reply libfprint/drivers/vfs7552_testing.c]: Fixed it now. I just used the header of your original file ;)

**Comment by yozoon** [comment]
I recently started working on the driver again. This time I began by implementing the libfprint2 API and reusing parts of your code. After a few days of work I can report that the driver works now and it even detects sufficient minutiae for the fingerprint verification. The key was to clean the image (and scale it by a factor of 4 to make life easier for the minutiae detection algorithm) before submitting it to libfprint - I achieved this by removing the sensor background (which is captured early in the await_finger_on sequence). You can find the code in this repo: https://github.com/yozoon/libfprint

**Comment by hmaarrfk** [comment]
pretty smart!

Did you do any interpolation?

**Comment by yozoon** [comment]
No, I just copied one pixel of the original image into four pixels of the scaled one - that seems to work just fine^^

**Comment by hmaarrfk** [comment]
hmm understood!

Are you going to try to upstream this?

**Comment by yozoon** [comment]
Yes, I would like to try that. Are you still using your XPS laptop? If so could you maybe test the driver on your system - I would be quite interested in whether the variance thresholds I defined are specific to my machine, or universally applicable.

**Comment by hmaarrfk** [comment]
My laptop's hinge was loose, and I needed to replace the bottom case.

I couldn't find one with the finger print sensor, so I got rid of that.

I can't really help on this front anymore!

**Comment by justus-hildebrand** [comment]
@yozoon I have zero experience on how to set up this project or with driver development in general, but I have a working XPS 2in1 running ElementaryOS (i think Juno, haven't been using it much lately) that I'd love to have a working fingerprint sensor on. So if you're still interested in this project and want me to help with the "trying it out on the right hardware" part, get in touch, I'd be happy to help! :)

**Comment by yozoon** [comment]
@justus-hildebrand first clone the repo:
```bash
git clone https://github.com/yozoon/libfprint.git
cd libfprint/
```
then initialize the meson build directory and specify that you only want to build the vfs7552 driver:
```bash
# Initialize the build directory
meson -Ddrivers=vfs7552 _build
# cd to newly created build directory
cd _build/
```
Build and install (but before that make sure that libfprint isn't installed on your system. Something like ```sudo apt remove libfprint``` should do the trick):
```bash
# Build
ninja
# Install
sudo ninja install
```
Now the driver is installed, but you'll have to install fprintd to actually use it for authentication:
```bash
# Install the libfprint daemon
sudo apt install fprintd
```
This provides some handy command line utilities for enrolling and verifying the fingerprint (I don't know if elementary OS has a native GUI for that like GNOME does)
```bash
# Enroll a fingerprint
fprintd-enroll
# Verify if the enrolled fingerprint is recognized
fprintd-verify
```
I used the following bash instruction during development to launch the fprintd-verify command multiple times, without having to re-enter the command every time:
```bash
while true ; do (fprintd-verify | grep "Verify result") ; done
```

To remove the library
```bash
# Uninstall
sudo ninja uninstall
```

Viel Erfolg! ;)

Cheers,
Julius

**Comment by justus-hildebrand** [comment]
when initializing the build dir, i get an error:
```
The Meson build system
Version: 0.49.2
Source dir: /home/justus/libfprint-fingerprint-test/libfprint
Build dir: /home/justus/libfprint-fingerprint-test/libfprint/_build
Build type: native build
Project name: libfprint
Project version: 1.90.2
Native C compiler: cc (gcc 7.4.0 "cc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
Native C++ compiler: c++ (gcc 7.4.0 "c++ (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0")
Build machine cpu family: x86_64
Build machine cpu: x86_64
Compiler for C supports arguments -Wall: YES
Compiler for C supports arguments -Wcast-align: YES
Compiler for C supports arguments -Wformat-nonliteral: YES
Compiler for C supports arguments -Wformat-security: YES
Compiler for C supports arguments -Wformat=2: YES
Compiler for C supports arguments -Wignored-qualifiers: YES
Compiler for C supports arguments -Wlogical-op: YES
Compiler for C supports arguments -Wmissing-declarations: YES
Compiler for C supports arguments -Wmissing-format-attribute: YES
Compiler for C supports arguments -Wmissing-include-dirs: YES
Compiler for C supports arguments -Wmissing-noreturn: YES
Compiler for C supports arguments -Wpointer-arith: YES
Compiler for C supports arguments -Wshadow: YES
Compiler for C supports arguments -Wtype-limits: YES
Compiler for C supports arguments -Wundef: YES
Compiler for C supports arguments -Wunused: YES
Compiler for C supports arguments -Werror=address: YES
Compiler for C supports arguments -Werror=array-bounds: YES
Compiler for C supports arguments -Werror=empty-body: YES
Compiler for C supports arguments -Werror=init-self: YES
Compiler for C supports arguments -Werror=int-to-pointer-cast: YES
Compiler for C supports arguments -Werror=main: YES
Compiler for C supports arguments -Werror=missing-braces: YES
Compiler for C supports arguments -Werror=nonnull: YES
Compiler for C supports arguments -Werror=redundant-decls: YES
Compiler for C supports arguments -Werror=return-type: YES
Compiler for C supports arguments -Werror=sequence-point: YES
Compiler for C supports arguments -Werror=trigraphs: YES
Compiler for C supports arguments -Werror=write-strings: YES
Compiler for C supports arguments -fno-strict-aliasing: YES
Compiler for C supports arguments -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_56: YES
Compiler for C supports arguments -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_56: YES
Compiler for C supports arguments -D_GNU_SOURCE: YES
Compiler for C supports arguments -DG_LOG_DOMAIN="libfprint": YES
Compiler for C supports arguments -Wimplicit-function-declaration: YES
Compiler for C supports arguments -Wmissing-prototypes: YES
Compiler for C supports arguments -Wnested-externs: YES
Compiler for C supports arguments -Wold-style-definition: YES
Compiler for C supports arguments -Wstrict-prototypes: YES
Compiler for C supports arguments -Werror=implicit: YES
Compiler for C supports arguments -Werror=pointer-to-int-cast: YES
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Dependency glib-2.0 found: YES 2.56.4
Dependency gio-unix-2.0 found: YES 2.56.4
Dependency gobject-2.0 found: YES 2.56.4
Found CMake: /usr/bin/cmake (3.10.2)
Dependency gusb found: NO (tried pkgconfig and cmake)

meson.build:83:0: ERROR:  Dependency "gusb" not found, tried pkgconfig and cmake

```
ive checked my installations through apt, i have `libgusb2` installed, which according to [this](https://packages.ubuntu.com/bionic/libgusb2) is the only gusb library in the ubuntu archives (elementary uses ubuntu package archives), but it does not seem to be the right dependency.

any idea where i could get the right lib?

**Comment by NICHOLAS85** [comment]
Make sure you are installing the `-dev` version of those packages. `libgusb-dev` should be what you're looking for. 

By the way I tried building this on Debian Sid but it seems the fprintd in repos doesn't work with this library.

**Comment by justus-hildebrand** [comment]
that did the trick, thanks @NICHOLAS85! @yozoon the lib finished building and let me install it, but I got a syntax error here:
```
[87/93] Generating FPrint-2.0.gir with a custom command.
/usr/include/stdlib.h:140: syntax error, unexpected identifier, expecting ',' or ';' in '# 140 "/usr/include/stdlib.h" 3 4' at 'strtof32'
/usr/include/stdlib.h:141: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '     char **__restrict __endptr)' at 'char'
/usr/include/stdlib.h:141: syntax error, unexpected ')', expecting ',' or ';' in '     char **__restrict __endptr)' at ')'
/usr/include/stdlib.h:147: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '     char **__restrict __endptr)' at 'char'
/usr/include/stdlib.h:147: syntax error, unexpected ')', expecting ',' or ';' in '     char **__restrict __endptr)' at ')'
/usr/include/stdlib.h:158: syntax error, unexpected identifier, expecting ',' or ';' in 'extern _Float32x strtof32x (const char *__restrict __nptr,' at 'strtof32x'
/usr/include/stdlib.h:159: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '       char **__restrict __endptr)' at 'char'
/usr/include/stdlib.h:159: syntax error, unexpected ')', expecting ',' or ';' in '       char **__restrict __endptr)' at ')'
/usr/include/stdlib.h:165: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '       char **__restrict __endptr)' at 'char'
/usr/include/stdlib.h:165: syntax error, unexpected ')', expecting ',' or ';' in '       char **__restrict __endptr)' at ')'
/usr/include/stdlib.h:233: syntax error, unexpected identifier in '         _Float32 __f)' at '_Float32'
/usr/include/stdlib.h:239: syntax error, unexpected identifier in '         _Float64 __f)' at '_Float64'
/usr/include/stdlib.h:251: syntax error, unexpected identifier in '   _Float32x __f)' at '_Float32x'
/usr/include/stdlib.h:257: syntax error, unexpected identifier in '   _Float64x __f)' at '_Float64x'
/usr/include/stdlib.h:316: syntax error, unexpected identifier, expecting ',' or ';' in '# 316 "/usr/include/stdlib.h" 3 4' at 'strtof32_l'
/usr/include/stdlib.h:317: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '       char **__restrict __endptr,' at 'char'
/usr/include/stdlib.h:318: syntax error, unexpected typedef-name, expecting identifier or '(' or '*' in '       locale_t __loc)' at 'locale_t'
/usr/include/stdlib.h:324: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '       char **__restrict __endptr,' at 'char'
/usr/include/stdlib.h:325: syntax error, unexpected typedef-name, expecting identifier or '(' or '*' in '       locale_t __loc)' at 'locale_t'
/usr/include/stdlib.h:337: syntax error, unexpected identifier, expecting ',' or ';' in 'extern _Float32x strtof32x_l (const char *__restrict __nptr,' at 'strtof32x_l'
/usr/include/stdlib.h:338: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '         char **__restrict __endptr,' at 'char'
/usr/include/stdlib.h:339: syntax error, unexpected typedef-name, expecting identifier or '(' or '*' in '         locale_t __loc)' at 'locale_t'
/usr/include/stdlib.h:345: syntax error, unexpected CHAR, expecting identifier or '(' or '*' in '         char **__restrict __endptr,' at 'char'
/usr/include/stdlib.h:346: syntax error, unexpected typedef-name, expecting identifier or '(' or '*' in '         locale_t __loc)' at 'locale_t'
g-ir-scanner: link: cc -o /home/justus/libfprint-fingerprint-test/libfprint/_build/tmp-introspectva2vmg5c/FPrint-2.0 /home/justus/libfprint-fingerprint-test/libfprint/_build/tmp-introspectva2vmg5c/FPrint-2.0.o -L. -Wl,-rpath,. -Wl,--no-as-needed -lfprint-2 -lfprint-2 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lgusb -lusb-1.0 -lm -L/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -Wl,-rpath,/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -L/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -Wl,-rpath,/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -L/home/justus/libfprint-fingerprint-test/libfprint
 /_build/libfprint -Wl,-rpath,/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -L/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -Wl,-rpath,/home/justus/libfprint-fingerprint-test/libfprint/_build/libfprint -lgio-2.0 -lgobject-2.0 -Wl,--export-dynamic -lgmodule-2.0 -pthread -lglib-2.0
[93/93] Linking target tests/test-fpi-device.
```
oddly enough that's the default stdlib.h, which I of course never touched. Am I using the wrong compiler or something?

Nonetheless i tried installing, which didn't error. got a whole bunch of `Warning: multiple "IDs" for constraint linkend: FpImage.` warnings, those look docs related though.

So I think it should be installed, however `fprintd-enroll` returns `list_devices failed: No devices available`, seems like it is not working as of now. Any ideas?

**Comment by yozoon** [comment]
@hmaarrfk my version of the driver just got merged into upstream libfprint: https://gitlab.freedesktop.org/libfprint/libfprint https://gitlab.freedesktop.org/libfprint/libfprint/-/merge_requests/178
So thanks again for your great work reverse-engineering the communication as well as writing the first version of the driver - now everyone owning an XPS from 2017 can use the fingerprint sensor on linux :)

**Comment by hmaarrfk** [comment]
Whoa. This is great news. Happy to see that open source can bring collaboration on this scale!!!!


I really think you did a lot of heavy lifting getting this past my initial research stage prototype. Personally, i had to let my fingerprint sensor go when my chassis broke and i had to replace it with one without the fingerprint sensor. Sad that I don't get to try the fruits of your labor.

**Pull request closed without merging**